File: server/providers/air_quality.py
"""Air quality via Open-Meteo — keyless. US AQI + key pollutants for the resolved
location (or widget-supplied lat/lon)."""

from __future__ import annotations

from typing import Any

import httpx

from ..shared.geo import get_location
from ..shared.providers import Provider, register


def _category(aqi: float | None) -> str:
    if aqi is None:
        return "—"
    for hi, label in [(50, "Good"), (100, "Moderate"), (150, "Unhealthy (sensitive)"),
                      (200, "Unhealthy"), (300, "Very unhealthy")]:
        if aqi <= hi:
            return label
    return "Hazardous"


class AirQualityProvider(Provider):
    name = "air-quality"
    ttl = 1800.0  # 30 min

    async def fetch(self, params: dict[str, Any]) -> dict[str, Any]:
        loc = await get_location()
        lat = float(params.get("lat") or loc["lat"])
        lon = float(params.get("lon") or loc["lon"])
        async with httpx.AsyncClient(timeout=8.0) as client:
            r = await client.get(
                "https://air-quality-api.open-meteo.com/v1/air-quality",
                params={
                    "latitude": lat,
                    "longitude": lon,
                    "current": "us_aqi,pm2_5,pm10,ozone,nitrogen_dioxide",
                    "timezone": "auto",
                },
            )
            r.raise_for_status()
            cur = r.json().get("current", {})
        aqi = cur.get("us_aqi")
        return {
            "location": {"city": loc.get("city"), "region": loc.get("region")},
            "aqi": aqi,
            "category": _category(aqi),
            "pollutants": {
                "pm2_5": cur.get("pm2_5"),
                "pm10": cur.get("pm10"),
                "ozone": cur.get("ozone"),
                "no2": cur.get("nitrogen_dioxide"),
            },
        }


register(AirQualityProvider())

File: web/js/widgets/air-quality.js
// Air Quality (US AQI) — keyless via Open-Meteo, uses home/IP geolocation
// unless lat/lon are set on the widget.
import { define } from "./registry.js";
import { el, fetchData, fmtNum } from "./dom.js";

function level(aqi) {
  if (aqi == null) return "na";
  if (aqi <= 50) return "good";
  if (aqi <= 100) return "moderate";
  if (aqi <= 150) return "usg";
  if (aqi <= 200) return "unhealthy";
  if (aqi <= 300) return "vunhealthy";
  return "hazardous";
}

function clean(o) {
  return Object.fromEntries(Object.entries(o).filter(([, v]) => v != null && v !== ""));
}

define("air-quality", {
  meta: { label: "Air quality", description: "US AQI + pollutants", category: "data" },
  schema: {
    fields: [
      { key: "lat", label: "Latitude (blank = home/auto)", type: "number" },
      { key: "lon", label: "Longitude (blank = home/auto)", type: "number" },
      { key: "showNo2", label: "Show NO₂", type: "boolean", default: false },
      { key: "cacheTtlSeconds", label: "Server cache TTL seconds (blank = default)", type: "number" },
    ],
  },
  async mount(root, widget) {
    const body = el("div", { class: "aqi" });
    root.appendChild(body);
    const handle = { body, widget };
    await this.refresh(handle);
    return handle;
  },
  async refresh(handle) {
    const s = handle.widget.settings || {};
    try {
      const d = await fetchData("air-quality", clean({
        lat: s.lat, lon: s.lon, cacheTtl: s.cacheTtlSeconds,
      }));
      const p = d.pollutants || {};
      const pollutants = [
        el("span", {}, `PM2.5 ${fmtNum(p.pm2_5, 0)}`),
        el("span", {}, `PM10 ${fmtNum(p.pm10, 0)}`),
        el("span", {}, `O₃ ${fmtNum(p.ozone, 0)}`),
      ];
      if (s.showNo2) pollutants.push(el("span", {}, `NO₂ ${fmtNum(p.no2, 0)}`));
      handle.body.replaceChildren(
        el("div", { class: "aqi-gauge " + level(d.aqi) }, [
          el("div", { class: "aqi-value" }, d.aqi == null ? "—" : fmtNum(d.aqi, 0)),
          el("div", { class: "aqi-cat" }, d.category || "—"),
        ]),
        el("div", { class: "aqi-sub" }, pollutants),
        el("div", { class: "aqi-loc" }, [d.location?.city, d.location?.region].filter(Boolean).join(", "))
      );
    } catch {
      handle.body.replaceChildren(el("div", { class: "widget-error" }, "air quality unavailable"));
    }
  },
});
